File: sdk/src/storage.ts
/**
 * @module Storage client implementation for cloud storage bucket operations with file upload, download, deletion, listing, and signed URL generation.
 */
import type { StorageObject } from "./types";
import {
  encodePathSegment,
  encodePathWithSlashes,
  normalizeStorageListResponse,
  normalizeStorageObject,
} from "./helpers";

interface StorageClientRuntime {
  request<T>(path: string, init?: RequestInit & { skipAuth?: boolean }): Promise<T>;
  getBaseURL(): string;
}

/**
 * Client for managing files in cloud storage buckets. Supports uploading, downloading, deleting, listing files, and generating signed URLs for time-limited file access.
 */
export class StorageClient {
  constructor(private client: StorageClientRuntime) {}

  private bucketPath(bucket: string): string {
    return `/api/storage/${encodePathSegment(bucket)}`;
  }

  private objectPath(bucket: string, name: string): string {
    return `${this.bucketPath(bucket)}/${encodePathWithSlashes(name)}`;
  }

  /** Upload a file to a bucket. */
  async upload(
    bucket: string,
    file: Blob | File,
    name?: string,
  ): Promise<StorageObject> {
    const form = new FormData();
    const resolvedName = name ?? (file instanceof File ? file.name : "upload");
    // Set the name as a separate form field so the server uses it directly.
    // FormData filename sanitization strips path separators (e.g. "dir/file.txt"
    // becomes "file.txt"), so names with "/" must be sent as a field, not just
    // as the Content-Disposition filename.
    form.append("name", resolvedName);
    form.append("file", file, resolvedName);
    const response = await this.client.request<StorageObject>(this.bucketPath(bucket), {
      method: "POST",
      body: form,
    });
    return normalizeStorageObject(response);
  }

  /** Get a download URL for a file. */
  downloadURL(bucket: string, name: string): string {
    return `${this.client.getBaseURL()}${this.objectPath(bucket, name)}`;
  }

  /** Delete a file from a bucket. */
  async delete(bucket: string, name: string): Promise<void> {
    return this.client.request(this.objectPath(bucket, name), {
      method: "DELETE",
    });
  }

  /** List files in a bucket. */
  async list(
    bucket: string,
    params?: { prefix?: string; limit?: number; offset?: number },
  ): Promise<{ items: StorageObject[]; totalItems: number }> {
    const qs = new URLSearchParams();
    if (params?.prefix) qs.set("prefix", params.prefix);
    if (params?.limit != null) qs.set("limit", String(params.limit));
    if (params?.offset != null) qs.set("offset", String(params.offset));
    const suffix = qs.toString() ? `?${qs}` : "";
    const response = await this.client.request<{ items: StorageObject[]; totalItems: number }>(
      `${this.bucketPath(bucket)}${suffix}`,
    );
    return normalizeStorageListResponse(response);
  }

  /** Get a signed URL for time-limited access to a file. */
  async getSignedURL(
    bucket: string,
    name: string,
    expiresIn?: number,
  ): Promise<{ url: string }> {
    return this.client.request(`${this.objectPath(bucket, name)}/sign`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ expiresIn: expiresIn ?? 3600 }),
    });
  }
}

File: sdk_python/src/allyourbase/storage.py

from __future__ import annotations

from typing import TYPE_CHECKING, Any, BinaryIO, Dict, Optional, Union
from urllib.parse import urlencode

from allyourbase.types import StorageListResponse, StorageObject

if TYPE_CHECKING:
    from allyourbase.client import AYBClient


class StorageClient:
    """Handles file storage operations."""

    def __init__(self, client: AYBClient) -> None:
        self._client = client

    @staticmethod
    def _build_path(path: str, params: Dict[str, str]) -> str:
        if not params:
            return path
        return f"{path}?{urlencode(params)}"

    async def upload(
        self,
        bucket: str,
        file: Union[bytes, BinaryIO],
        name: str,
        *,
        content_type: Optional[str] = None,
    ) -> StorageObject:
        url = f"{self._client.base_url}/api/storage/{bucket}"
        headers: Dict[str, str] = {}
        if self._client.token is not None:
            headers["Authorization"] = f"Bearer {self._client.token}"

        file_data = file if isinstance(file, bytes) else file.read()
        files = {"file": (name, file_data, content_type or "application/octet-stream")}

        resp = await self._client._http.post(url, headers=headers, files=files)

        if resp.status_code < 200 or resp.status_code >= 300:
            from allyourbase.client import _raise_error

            _raise_error(resp)

        return StorageObject.model_validate(resp.json())

    def download_url(self, bucket: str, name: str) -> str:
        return f"{self._client.base_url}/api/storage/{bucket}/{name}"

    async def delete(self, bucket: str, name: str) -> None:
        await self._client._request(
            f"/api/storage/{bucket}/{name}",
            method="DELETE",
        )

    async def list(
        self,
        bucket: str,
        *,
        prefix: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> StorageListResponse:
        params: Dict[str, str] = {}
        if prefix is not None:
            params["prefix"] = prefix
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)

        path = self._build_path(f"/api/storage/{bucket}", params)
        resp = await self._client._request(path)
        if resp is None:
            raise RuntimeError("Expected response body for storage list")
        return StorageListResponse.model_validate(resp.json())

    async def get_signed_url(
        self,
        bucket: str,
        name: str,
        *,
        expires_in: int = 3600,
    ) -> str:
        resp = await self._client._request(
            f"/api/storage/{bucket}/{name}/sign",
            method="POST",
            json={"expiresIn": expires_in},
        )
        if resp is None:
            raise RuntimeError("Expected response body for signed URL")
        result: Any = resp.json()
        if not isinstance(result, dict):
            raise RuntimeError("Expected object payload for signed URL")
        url = result.get("url")
        if not isinstance(url, str):
            raise RuntimeError("Expected signed URL string in response")
        return url
